### DISCUSSION TITLE
OIDCUnAutzAction - iframe/frame auto-detection mechanism override

### DISCUSSION THREAD

**Discussion by spiazzi** [OpenIDC/mod_auth_openidc#1121] [Q&A]
We're using mod_auth_openidc to protect different web applications in a single Apache VirtualHost; each application is published with a specific context-root and requires (through Require directive) a specific claim value in "id_token" in order to authorize the access.

Some of the webapps use iframes / framesets where other applications are loaded, so a solid SSO mechanisms between applications is required.

 

We're trying to implement a smooth SSO flow through the iframes, but we are facing 2 main issues:

1. **UnAuth user**: if user tries to refresh an iframe / frame after the mod_auth_openidc session is expired, a 401 error is returned instead of starting the OIDC authentication processas stated in the conf https://github.com/OpenIDC/mod_auth_openidc/blob/109c841d23ef5036db2dfa537e4bef0560749b43/auth_openidc.conf#L929, we managed to override the iframe/frame auto-detection mechanism by using the specific argument provided by "OIDCUnAuthAction" directive -> **SOLVED** -> OIDCUnAuthAction 401 "%{HTTP:X-Requested-With} == 'XMLHttpRequest' || ( -n %{HTTP:Sec-Fetch-Mode} && %{HTTP:Sec-Fetch-Mode} != 'navigate' ) || || ( -n %{HTTP:Sec-Fetch-Dest} && %{HTTP:Sec-Fetch-Dest} != 'iframe' && %{HTTP:Sec-Fetch-Dest} != 'frame' && %{HTTP:Sec-Fetch-Dest} != 'document') || ( ( %{HTTP_ACCEPT} !~ m#text/html# ) && ( %{HTTP_ACCEPT} !~ m#application/xhtml\+xml# ) && ( %{HTTP_ACCEPT} !~ m#\*/\*# ) )"
2. **UnAutz user:** we're trying to handle the authorization process through an OIDC "step-up" flow, but again, if the user tries to load in the iframe an application that is not already auhtorized to, a 401 error is returned instead of starting the OIDC step-up authorization process -> we wanted to use the same approach by overriding the iframe/frame auto-detection mechanism, but (reading the docs) we could not find the right way to specify the same override logic for OIDCUnAutzAction -> **ISSUE PENDING**

Is there a way to specifiy the iframe/frame auto-detection mechanism override for OIDCUnAutzAction like we did for the OIDCUnAuthAction?

 

If not already, can this possibility be added to the OIDCUnAutzAction too?

**Comment by zandbelt** [comment]
iframe based SSO is deemed insecure, hence the default settings and there are no plans to deviate from that; one should rather aim to handle authentication at the toplevel

**Reply by perry19987** [reply]
> iframe based SSO is deemed insecure, hence the default settings and there are no plans to deviate from that; one should rather aim to handle authentication at the toplevel

Hello @zandbelt 
I resume this thread cause i have a question raleted to this topic.

I also have a web portal that internally, through iframe, opens various applications, each protected by different authorization levels that therefore require a step-up authorization process to be accessed.

Due to restrictions on the use of iframes, we are working with directives related to the response header Content Security Policy (e.g., frame-ancestor and similar), which, in our opinion, mitigate the risks associated with the use of iframes, such as:
- explicitly allowing the portal to embed the protected app via iframe;
- preventing unauthorized embedding from other origins;
- potentially helping browsers treat the iframe as first-party in some contexts, depending on cookie settings and browser behavior.

I wanted to ask you if, under these conditions, it is still feasible to foresee the possibility of overriding the OIDCUnAutzAction directive, considering that the execution protection is guided by the indicated directive.

**Comment by zandbelt** [comment]
that is possible already since 2.4.16 see: https://github.com/OpenIDC/mod_auth_openidc/commit/0aa5431d917a635f77e473fdb725f4438c561b74 and the release notes at: https://github.com/OpenIDC/mod_auth_openidc/releases/tag/v2.4.16.1
